File: app/src/main/java/ericdiaz/program/currencyconveterlive2019/viewmodel/ExchangeRateViewModel.java
package ericdiaz.program.currencyconveterlive2019.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import javax.inject.Inject;

import ericdiaz.program.currencyconveterlive2019.repository.BaseRepository;
import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;

public class ExchangeRateViewModel extends BaseViewModel {

    private final MutableLiveData<State> exchangeRateData = new MutableLiveData<>();
    private final BaseRepository exchangeRateRepository;

    @Inject
    public ExchangeRateViewModel(@NonNull final BaseRepository exchangeRateRepository) {
        this.exchangeRateRepository = exchangeRateRepository;
    }

    public void getConversionValue(@NonNull final String data,
                                   @NonNull final String baseCurrency,
                                   String foreignCurrency,
                                   String baseCurrencyAmount) {
        addDisposables(
          exchangeRateRepository
            .requestExchangeRates(data, baseCurrency)
            .delaySubscription(
              Completable.fromAction(() -> exchangeRateData.setValue(State.Loading.INSTANCE)))
            .map(exchangeRateResponse -> {

                Double baseAmount = Double.valueOf(baseCurrencyAmount);
                Double conversionRate = exchangeRateResponse.getRatesMap().get(foreignCurrency);

                return conversionRate != null ? String.valueOf((baseAmount * conversionRate)) : "Error, value is null";
            })
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
              exchangeRateResponse -> exchangeRateData.setValue(new State.Success(exchangeRateResponse)),
              throwable -> exchangeRateData.setValue(new State.Failure(throwable))));
    }

    public LiveData<State> getExchangeRateData() {
        return exchangeRateData;
    }
}

File: app/src/main/java/ericdiaz/program/currencyconveterlive2019/viewmodel/ExchangeRateViewModel.kt
package ericdiaz.program.currencyconveterlive2019.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import ericdiaz.program.currencyconveterlive2019.extensions.getExchangeValue
import ericdiaz.program.data.model.ExchangeRateResponse
import ericdiaz.program.data.repository.ExchangeRateDatabaseRepository
import ericdiaz.program.data.repository.ExchangeRateNetworkRepository
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy

class ExchangeRateViewModel(private val exchangeRateNetworkRepository: ExchangeRateNetworkRepository,
                            private val exchangeRateDatabaseRepository: ExchangeRateDatabaseRepository) : BaseViewModel() {

    private val exchangeRateData = MutableLiveData<State>()
    private val currencyProfilesData = MutableLiveData<State>()
    lateinit var baseCurrency: String
    lateinit var foreignCurrency: String
    lateinit var baseCurrencyAmount: String

    fun getConversionValue(date: String = "latest") {
        if (baseCurrencyAmount != "0.00") {
            addDisposables(
                    exchangeRateDatabaseRepository
                            .requestExchangeRates(date, baseCurrency)

                            .delaySubscription(
                                    Completable.fromAction { exchangeRateData.setValue(State.Loading) }
                            )

                            .onErrorResumeNext {
                                exchangeRateNetworkRepository.requestExchangeRates(date, baseCurrency)
                                        .doOnSuccess { exchangeRateResponse ->
                                            exchangeRateResponse?.run {
                                                exchangeRateDatabaseRepository.insertExchangeRateResponse(exchangeRateResponse)
                                            }
                                        }
                            }

                            .map { (baseCurrency, ratesMap, date): ExchangeRateResponse ->

                                val conversionRate = ratesMap[foreignCurrency]

                                val conversionValue = conversionRate?.getExchangeValue(baseCurrencyAmount)
                                        ?: "Error, value is null"

                                State.Success(
                                        conversionValue,
                                        " (1) $baseCurrency = ($conversionRate) $foreignCurrency",
                                        "Rates as of : $date")
                            }

                            .observeOn(AndroidSchedulers.mainThread())

                            .subscribeBy(
                                    onSuccess = { state -> exchangeRateData.value = state },
                                    onError = { throwable -> exchangeRateData.value = State.Failure(throwable) }
                            )
            )
        } else {
            exchangeRateData.value = State.Zero
        }
    }

    fun getCurrencyProfiles() {
        addDisposables(exchangeRateDatabaseRepository
                .requestCurrencyProfiles()
                .onErrorResumeNext {
                    exchangeRateNetworkRepository
                            .requestCurrencyProfiles()
                            .doOnSuccess {
                                exchangeRateDatabaseRepository.insertCurrencyProfileMap(it)
                            }.retry(3)
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onSuccess = { response -> currencyProfilesData.value = State.Success(currencyProfileMap = response) },
                        onError = { throwable -> currencyProfilesData.value = State.Failure(throwable) }
                ))
    }

    fun getExchangeRateData(): LiveData<State> {
        return exchangeRateData
    }

    fun getCurrencyProfilesData(): LiveData<State> {
        return currencyProfilesData
    }
}
